File: Android/app/src/main/java/live/ditto/pos/core/data/repository/SaleItemsRepository.kt
package live.ditto.pos.core.data.repository

import com.ditto.kotlin.Ditto
import com.ditto.kotlin.DittoSyncSubscription
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import live.ditto.ditto_wrapper.DittoManager
import live.ditto.pos.core.data.SaleItem
import live.ditto.pos.core.data.observeAsFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SaleItemsRepository @Inject constructor(
    private val dittoManager: DittoManager,
    private val locationsRepository: LocationsRepository
) {
    private val ditto: Ditto get() = dittoManager.requireDitto()
    private var subscription: DittoSyncSubscription? = null
    private val repoScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    init {
        locationsRepository.locationIdFlow()
            .filter { it.isNotEmpty() }
            .distinctUntilChanged()
            .onEach { locationId -> setActiveLocation(locationId) }
            .launchIn(repoScope)
    }

    private fun setActiveLocation(locationId: String) {
        subscription?.close()
        // Subscription = the sync set: every sale item for this location.
        // ORDER BY / LIMIT are illegal on a subscription in v5 — ordering is a
        // presentation concern and lives on the observer in observeLocationSaleItems.
        subscription = try {
            ditto.sync.registerSubscription(
                """
                    SELECT * FROM ${SaleItem.COLLECTION_NAME}
                    WHERE _id.locationId = :locationId
                """.trimIndent(),
                mapOf("locationId" to locationId)
            )
        } catch (error: Throwable) {
            reportSubscriptionFailure("subscribe sale_items", error)
        }
    }

    fun observeLocationSaleItems(locationId: String): Flow<List<SaleItem>> =
        ditto.store.observeAsFlow(
            query = """
                SELECT * FROM ${SaleItem.COLLECTION_NAME}
                WHERE _id.locationId = :locationId
                ORDER BY name
            """.trimIndent(),
            args = mapOf("locationId" to locationId)
        )
}

File: iOS/DittoPOS/Data/Repositories/SaleItemsRepository.swift
//
//  SaleItemsRepository.swift
//  DittoPOS
//
//  Copyright © 2026 DittoLive Incorporated. All rights reserved.
//

import Combine
import DittoSwift

/// Owns the `sale_items` collection: registers the per-location subscription
/// and exposes the synced menu items to consumers. Mirrors Android's
/// `SaleItemsRepository`.
@MainActor final class SaleItemsRepository: ObservableObject {

    @Published private(set) var locationSaleItems: [SaleItem] = []

    private let dittoManager: DittoManager
    private var ditto: Ditto { dittoManager.ditto }
    private var store: DittoStore { ditto.store }
    private var sync: DittoSync { ditto.sync }

    private var subscription: DittoSyncSubscription?
    private var observer = AnyCancellable({})
    private var locationSubscription = AnyCancellable({})

    init(dittoManager: DittoManager = .shared, locationsRepository: LocationsRepository) {
        self.dittoManager = dittoManager
        locationSubscription = locationsRepository.$currentLocationId
            .compactMap { $0 }
            .removeDuplicates()
            .sink { [weak self] locationId in
                self?.setActiveLocation(locationId)
            }
    }

    private func setActiveLocation(_ locationId: String) {
        subscription?.cancel()

        // Subscription = the sync set: every sale item for this location. ORDER BY
        // / LIMIT are illegal on a subscription in v5. A failed registration must
        // NOT skip the observer below — it's a local read over the store,
        // independent of the sync set (matches Android's separation).
        do {
            subscription = try sync.registerSubscription(
                query: """
                    SELECT * FROM \(SaleItem.collectionName)
                    WHERE _id.locationId = :locationId
                    """,
                arguments: ["locationId": locationId]
            )
        } catch {
            reportSubscriptionFailure("subscribe sale_items", error)
        }

        // Observer = the local read that drives the UI, ordered for display.
        // Its own query string, independent of the subscription's success.
        observer = store.observePublisher(
            query: """
                SELECT * FROM \(SaleItem.collectionName)
                WHERE _id.locationId = :locationId
                ORDER BY name
                """,
            arguments: ["locationId": locationId],
            mapTo: SaleItem.self
        )
        .replaceError(with: [])
        .assign(to: \.locationSaleItems, on: self)
    }
}
